smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,365 @@
+                f"exactly `{self.launcher_type}`"
+            )
+
+        def format_(exe: ExecutableProtocol, env: _EnvironMappingType) -> _LaunchableT:
+            return self.formatter(settings, exe, env)
+
+        return _LauncherAdapter(launcher, format_)
+
+    def configure_first_compatible_launcher(
+        self,
+        with_settings: _DispatchableT,
+        from_available_launchers: t.Iterable[LauncherProtocol[t.Any]],
+    ) -> _LaunchConfigType:
+        """Configure the first compatible adapter launch to launch with the
+        provided settings. Launchers are iterated and discarded from the
+        iterator until the iterator is exhausted.
+
+        :param with_settings: The settings with which to configure the launcher
+        :param from_available_launchers: An iterable that yields launcher instances
+        :raises errors.LauncherNotFoundError: No compatible launcher was
+            yielded from the provided iterator.
+        :returns: A launcher configured with the provided settings.
+        """
+        launcher = helpers.first(self._is_compatible_launcher, from_available_launchers)
+        if launcher is None:
+            raise errors.LauncherNotFoundError(
+                f"No launcher of exactly type `{self.launcher_type.__name__}` "
+                "could be found from provided launchers"
+            )
+        return self.create_adapter_from_launcher(launcher, with_settings)
+
+
+@t.final
+class _LauncherAdapter(t.Generic[Unpack[_Ts]]):
+    """An adapter class that will wrap a launcher and allow for a unique
+    signature of for its `start` method
+    """
+
+    def __init__(
+        self,
+        launcher: LauncherProtocol[_LaunchableT],
+        map_: t.Callable[[Unpack[_Ts]], _LaunchableT],
+    ) -> None:
+        """Initialize a launcher adapter
+
+        :param launcher: The launcher instance this class should wrap
+        :param map_: A callable with arguments for the new `start` method that
+            can translate them into the expected launching type for the wrapped
+            launcher.
+        """
+        # NOTE: We need to cast off the `_LaunchableT` -> `Any` in the
+        #       `__init__` method signature to hide the transform from users of
+        #       this class. If possible, this type should not be exposed to
+        #       users of this class!
+        self._adapt: t.Callable[[Unpack[_Ts]], t.Any] = map_
+        self._adapted_launcher: LauncherProtocol[t.Any] = launcher
+
+    def start(self, *args: Unpack[_Ts]) -> LaunchedJobID:
+        """Start a new job through the wrapped launcher using the custom
+        `start` signature
+
+        :param args: The custom start arguments
+        :returns: The launched job id provided by the wrapped launcher
+        """
+        payload = self._adapt(*args)
+        return self._adapted_launcher.start(payload)
+
+
+DEFAULT_DISPATCHER: t.Final = Dispatcher()
```

Comment:
Mayhaps add a short doc string here for line 301 and 305

Reply:
Added!